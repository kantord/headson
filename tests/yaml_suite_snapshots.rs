use assert_cmd::Command;
use insta::assert_snapshot;
use std::fs;
use std::path::Path;
use test_each_file::test_each_path;

fn run_cli_yaml_with_budget(input: &[u8], budget: usize) -> String {
    let budget_s = budget.to_string();
    let assert = Command::cargo_bin("headson")
        .unwrap()
        .args([
            "--no-color",
            "-n",
            &budget_s,
            "--string-cap",
            "1000000",
            "-f",
            "yaml",
            "-i",
            "yaml",
        ])
        .write_stdin(input)
        .assert()
        .success();
    String::from_utf8_lossy(&assert.get_output().stdout).into_owned()
}

fn is_yaml_file(path: &Path) -> bool {
    path.extension().map(|e| e == "yaml").unwrap_or(false)
}

fn stem_str(path: &Path) -> String {
    path.file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("unknown")
        .to_string()
}

const BUDGET_TIGHT: usize = 80; // significantly truncated but readable
const BUDGET_MED: usize = 400; // slightly truncated for many
const BUDGET_FULL: usize = 1_000_000; // effectively untruncated

test_each_path! { in "tests/fixtures/yaml/yaml-test-suite" => yaml_snapshot_case }

fn yaml_snapshot_case(path: &Path) {
    if !is_yaml_file(path) {
        return;
    }
    let input = fs::read(path).expect("read yaml");
    let name = stem_str(path);

    // Tight budget
    let out_tight = run_cli_yaml_with_budget(&input, BUDGET_TIGHT);
    assert_snapshot!(format!("yaml_suite_{}_tight", name), out_tight);

    // Medium budget
    let out_med = run_cli_yaml_with_budget(&input, BUDGET_MED);
    assert_snapshot!(format!("yaml_suite_{}_med", name), out_med);

    // Full budget
    let out_full = run_cli_yaml_with_budget(&input, BUDGET_FULL);
    assert_snapshot!(format!("yaml_suite_{}_full", name), out_full);
}

// No output normalization: runtime behavior is deterministic (aliases -> "*alias").
