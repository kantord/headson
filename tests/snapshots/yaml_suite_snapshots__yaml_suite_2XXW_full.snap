---
source: tests/yaml_suite_snapshots.rs
expression: out_full
---
- dump: "--- !!set\nMark McGwire:\nSammy Sosa:\nKen Griff:\n"
  from: "http://www.yaml.org/spec/1.2/spec.html#id2761758"
  json: "{\n  \"Mark McGwire\": null,\n  \"Sammy Sosa\": null,\n  \"Ken Griff\": null\n}\n"
  name: Spec Example 2.25. Unordered Sets
  tags: spec mapping unknown-tag explicit-key
  tree: "+STR\n +DOC ---\n  +MAP <tag:yaml.org,2002:set>\n   =VAL :Mark McGwire\n   =VAL :\n   =VAL :Sammy Sosa\n   =VAL :\n   =VAL :Ken Griff\n   =VAL :\n  -MAP\n -DOC\n-STR\n"
  yaml: "# Sets are represented as a\n# Mapping where each key is\n# associated with a null value\n--- !!set\n? Mark McGwire\n? Sammy Sosa\n? Ken Griff\n"
